### ISSUE TITLE
Receipts Root Mismatch

### ISSUE THREAD

**Issue by jflo** [hyperledger/besu#3725]
Investigation Ongoing

### Description
- The besu/teku pair halted at block 14564207 (TTD block)
    - Thats the TTD block. Besu used it as the common ancestor with peers it gathered.
    - Besu started a backward sync, before receiving its first fcu.
 ```
2022-04-12 14:58:01.206+00:00 | vert.x-worker-thread-0 | INFO  | MergeCoordinator | appending block hash 0x0ef93f8924634cc54164f73dd1e5ccb1a33f3802f2f4071ef781067e4aaca271 to backward sync
```

- that has would end up being the first one Besu reported as not being able to find the header for:

```
2022-04-12 14:58:36.395+00:00 | nioEventLoopGroup-3-8 | WARN  | BackwardSyncContext | there was an uncaught exception during backward sync
java.util.concurrent.CompletionException: org.hyperledger.besu.ethereum.eth.sync.backwardsync.BackwardSyncException: Did not receive a header for hash {}0x0ef93f8924634cc54164f73dd1e5ccb1a33f3802f2f4071ef781067e4aaca271
```
- Is the completion exception the result of a timeout on trying to find that hash from a peer?

```
2022-04-12 14:58:42.606+00:00 | EthScheduler-Timer-0 | ERROR | BackwardSyncLookupService | Failed to fetch blocks because java.util.concurrent.TimeoutException: Timeout after 5000 MILLISECONDS Current peers: 12.  Waiting for few seconds ...
```

```
2022-04-12 21:20:38.008+00:00 | nioEventLoopGroup-3-8 | INFO  | BackwardSyncPhase | Backward Phase finished.
```

- Still getting messages of blocks being appended to backward sync

```
2022-04-12 21:21:52.780+00:00 | ForkJoinPool.commonPool-worker-19 | WARN  | MainnetBlockBodyValidator | Invalid block: receipts root mismatch (expected=0x93462180b146c740623116abbb976111c9c4ac3f6320ab0b0591ae65c1b6c073, actual=0x2f68c64cd2072bb7b19e5fa67
8940a5c26219ec7f6b4d72469985af1a29503b1)
2022-04-12 21:21:52.781+00:00 | ForkJoinPool.commonPool-worker-19 | ERROR | MainnetBlockValidator | Block body not valid. Block 14564235 (0x6bd5f7eaf8517616ab4cf4443eb19f451caa1f115ee39157baa035b7bb3afeee)
```

- About 6 hours later, Besu moved forward to block 14564234
    - Probably because it found peers. Could confirm this with prometheus metrics.
- Once it got to 14564235 (0x6bd5f7eaf8517616ab4cf4443eb19f451caa1f115ee39157baa035b7bb3afeee) it halted due to a receipts root mismatch. We are waiting on a block trace from Geth, which passed this block 
(14564235 (0xDE3B8B),  hash 0x6bd5f7eaf8517616ab4cf4443eb19f451caa1f115ee39157baa035b7bb3afeee) just fine.

**Comment by garyschulte** [comment]
besu's trace of that (perceived bad) block:

[trace-0xde3b8b.tar.gz](https://github.com/hyperledger/besu/files/8477851/trace-0xde3b8b.tar.gz)

**Comment by MariusVanDerWijden** [comment]
Geth trace: 
```
cat block_0x6bd5f7ea-34-0x9a16559f-2615258477 
{"pc":0,"op":96,"gas":"0x10d88","gasCost":"0x3","memSize":0,"stack":[],"depth":1,"refund":0,"opName":"PUSH1"}
{"pc":2,"op":96,"gas":"0x10d85","gasCost":"0x3","memSize":0,"stack":["0x80"],"depth":1,"refund":0,"opName":"PUSH1"}
{"pc":4,"op":82,"gas":"0x10d82","gasCost":"0xc","memSize":96,"stack":["0x80","0x40"],"depth":1,"refund":0,"opName":"MSTORE"}
{"pc":5,"op":52,"gas":"0x10d76","gasCost":"0x2","memSize":96,"stack":[],"depth":1,"refund":0,"opName":"CALLVALUE"}
{"pc":6,"op":128,"gas":"0x10d74","gasCost":"0x3","memSize":96,"stack":["0xc2db6b71da9000"],"depth":1,"refund":0,"opName":"DUP1"}
{"pc":7,"op":21,"gas":"0x10d71","gasCost":"0x3","memSize":96,"stack":["0xc2db6b71da9000","0xc2db6b71da9000"],"depth":1,"refund":0,"opName":"ISZERO"}
{"pc":8,"op":97,"gas":"0x10d6e","gasCost":"0x3","memSize":96,"stack":["0xc2db6b71da9000","0x0"],"depth":1,"refund":0,"opName":"PUSH2"}
{"pc":11,"op":87,"gas":"0x10d6b","gasCost":"0xa","memSize":96,"stack":["0xc2db6b71da9000","0x0","0x10"],"depth":1,"refund":0,"opName":"JUMPI"}
{"pc":12,"op":96,"gas":"0x10d61","gasCost":"0x3","memSize":96,"stack":["0xc2db6b71da9000"],"depth":1,"refund":0,"opName":"PUSH1"}
{"pc":14,"op":128,"gas":"0x10d5e","gasCost":"0x3","memSize":96,"stack":["0xc2db6b71da9000","0x0"],"depth":1,"refund":0,"opName":"DUP1"}
{"pc":15,"op":253,"gas":"0x10d5b","gasCost":"0x0","memSize":96,"stack":["0xc2db6b71da9000","0x0","0x0"],"depth":1,"refund":0,"opName":"REVERT"}
{"pc":15,"op":253,"gas":"0x10d5b","gasCost":"0x0","memSize":96,"stack":["0xc2db6b71da9000"],"depth":1,"refund":0,"opName":"REVERT","error":"execution reverted"}
{"output":"","gasUsed":"0x2d","time":77271,"error":"execution reverted"}
```

**Comment by jflo** [comment]
geths trace of the same block
[geth-trace.tar.gz](https://github.com/hyperledger/besu/files/8492879/geth-trace.tar.gz)

**Comment by jflo** [comment]
Transaction 30 in the block, at pc 5331 the stack differs between the two. The bottom value of the stack in 0x0 in geth, and 0x87e795ecd14675853ef0be2fd164477e6f8910cf in besu. Investigation continues.

**Comment by matkt** [comment]
transaction 0 on Besu seems to fail and consume more than Geth. 

Besu : 
> {"pc":0,"op":0,"gas":"0x834","gasCost":"0x0","memory":"0x","memSize":0,"stack":[],"returnData":null,"depth":1,"refund":0,"opName":"STOP","error":""}
{"output":"","gasUsed":"0x5208","time":3370004}

Geth :
> {"pc":0,"op":96,"gas":"0xdef3c","gasCost":"0x3","memSize":0,"stack":[],"depth":1,"refund":0,"opName":"PUSH1"}
{"pc":2,"op":96,"gas":"0xdef39","gasCost":"0x3","memSize":0,"stack":["0x80"],"depth":1,"refund":0,"opName":"PUSH1"}
{"pc":4,"op":82,"gas":"0xdef36","gasCost":"0xc","memSize":96,"stack":["0x80","0x40"],"depth":1,"refund":0,"opName":"MSTORE"}
{"pc":5,"op":96,"gas":"0xdef2a","gasCost":"0x3","memSize":96,"stack":[],"depth":1,"refund":0,"opName":"PUSH1"}
{"pc":7,"op":54,"gas":"0xdef27","gasCost":"0x2","memSize":96,"stack":["0x4"],"depth":1,"refund":0,"opName":"CALLDATASIZE"}
{"pc":8,"op":16,"gas":"0xdef25","gasCost":"0x3","memSize":96,"stack":["0x4","0x24"],"depth":1,"refund":0,"opName":"LT"} 
...
{"output":"0000000000000000000000000000000000000000000000000000000000000001","gasUsed":"0x4cec","time":11398891}


it should be replayed in my opinion and understand why we do not go as far as Geth

**Comment by jflo** [comment]
Am so far unable to confirm that tx 30 is the one that is breaking the receipts root. Prior theories that the block traces were relevant, no longer hold water; I attached a debugger to the mainnet-shadowfork1 host, which is looping on the bad block. With a breakpoint before the error is emitted, I can evaluate all other methods in `MainnetBlockBodyValidator` used to validate the block:
```
validateStateRoot
validateTransactionsRoot
validateGasUsed
validateLogsBloom
validateEthHash
```
They all pass. Would love for someone to check my work, because if that is all correct, and our traces are also out of line with Geth, then the traces cannot be trusted. If besus traces were correct, state root should not have validated.

I've isolated where the receipts are hashed, and if I can get a list of receipts from Geth, I could compare them to find the problem ones. The hash is based on RLP encoded receipts though, and I haven't found a way to get those receipts from Geth in RLP, only in json.
